### PR TITLE
tkt-59247: Add Functional Tests

### PR DESCRIPTION
**Pull request by sonicaj**
This PR aims to add more tests to iocage.

FreeNAS Ticket: #59247

### CODE REVIEW

File: tests/data_classes.py [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+
+    def df_parse(self):
+        self.name, self.crt, self.res, self.qta, self.use, self.ava = \
+            self.standard_parse()
+
+    def snapshot_parse(self):
+        self.name, self.created, self.rsize, self.used = self.standard_parse()
+
+
+class ZFS:
```
Comment (skarekrow) [inline tests/data_classes.py:188]: Perhaps we shift to subprocess zfs again starting here as well? That way we can remove our dep on it sooner and easier.

File: tests/data_classes.py [hunk 2]
```diff
@@ -0,0 +1,647 @@
...
+                template = '-'
+
+        return Row({
+            'name': self.short_name if short_name else self.name,
+            'jid': self.jid,
+            'state': 'up' if self.running else 'down',
+            'boot': self.config.get('boot', 'off'),
+            'type': 'jail' if not self.is_template else 'template',
+            # TODO: Add support for plugins
+            'ip6': '-',  # FIXME: Change when ip6 tests are added
```
Comment (skarekrow) [inline tests/data_classes.py:378]: Use `'none'` to be like the other props that aren't specified
```suggestion
             'ip6': 'none',  # FIXME: Change when ip6 tests are added
```

File: tests/data_classes.py [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+    def is_rcjail(self):
+        return self.config.get('boot', 'off') == 'on'
+
+    @property
+    def ip(self):
+        assert self.running is True
+        try:
+            # TODO: we should probably load interface from config and
+            # make changes if necessary to it's value then
+            # TODO: Discuss with Brandon about support for older releases
```
Comment (skarekrow) [inline tests/data_classes.py:520]: Once again, what support do you want to discuss :P

File: tests/data_classes.py [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+
+            return pool['root_dataset']['properties']
+
+    def zfs_get_property(self, identifier, key):
+        try:
+            return self._zfs_get_properties(identifier)[key]['value']
+        except Exception:
+            return '-'
+
+    def get_dataset_safely(self, dataset_name):
```
Comment (skarekrow) [inline tests/data_classes.py]: Is there ever a time we don't want to get them safely? Perhaps just remove the other method and rename this one.

**Review by skarekrow — CHANGES_REQUESTED** [review]

File: tests/data_classes.py (hunk 2 above)
Comment (sonicaj) [reply tests/data_classes.py:378]: Umm this is to ensure matching output with the listing of jails etc. We don't use this for any other reason. If we change it to none, then the comparison will fail

Comment (skarekrow) [reply tests/data_classes.py:378]: Got it :+1:

**Review by skarekrow — DISMISSED** [review]

**Review by skarekrow — APPROVED** [review]

**Pull request merged**